dashboard/config/libraries/variableLog.interpreted.js

```diff
@@ -1,78 +1,110 @@
 var studentVarToken = false;
-var varLog={};
-if(!validationProps){
-  var validationProps ={};
-}
+var previousVarLog;
 
-function updateLog(){
-  var start = Object.keys(window).indexOf("studentVarToken")+1;
-  var end = Object.keys(window).indexOf("__jsCB");
+// Create an array of objects representing within the variables that have been
+// declared or updated within the Blockly environment. All Blockly variables
+// are global, and we would like to find a more efficient way to find just
+// those used by the student.
+function buildVariableLog() {
+  text("Building variable log...",0,15);
+  var varLog = {};
+  var start = Object.keys(window).indexOf("studentVarToken") + 1;
+  var end = Object.keys(window).length;
   var index;
-  for(var i=start;i<end;i++){
-    if(window[Object.keys(window)[i]] != undefined){
-      if (typeof window[Object.keys(window)[i]] == "number" || typeof window[Object.keys(window)[i]] == "string"|| typeof window[Object.keys(window)[i]] == "boolean") {
-        varLog[Object.keys(window)[i]] = window[Object.keys(window)[i]];
-      }
+  for (var i = start; i < end; i++) {
+    // Blockly functions and behaviors should be excluded. 
+    if (
+      typeof window[Object.keys(window)[i]] == "number" ||
+      typeof window[Object.keys(window)[i]] == "string" ||
+      typeof window[Object.keys(window)[i]] == "boolean"
+    ) {
+      varLog[Object.keys(window)[i]] = window[Object.keys(window)[i]];
     }
   }
+  return varLog;
+}
 
-  varWatchers(varLog);
+function detectVariableLogChange() {
+  var result = false;
+  if(varLog && previousVarLog){
+    for (var property in varLog) {
+      if (varLog[property] !== previousVarLog[property]) {
+        //window[Object.keys(window)[Object.keys(window).indexOf(property)]]
+        result = true;
+      }
+    }
+  }
+  return result;
+}
 
-  if(validationProps){//is Sprite Lab
-    if(!validationProps.previous){
-      validationProps.previous = {};
+function updateVariableLog() {
+  for (var property in varLog) {
+    if (varLog[property] !== window[Object.keys(window)[Object.keys(window).indexOf(property)]]) {
+      varLog[property] = window[Object.keys(window)[Object.keys(window).indexOf(property)]];
     }
-    detectVarChange();
-    validationProps.previous.varLog=JSON.parse(JSON.stringify(varLog));
   }
+}
 
+function storeVariableLogforPrevious() {
+  previousVarLog = JSON.parse(JSON.stringify(varLog));
```

Comment:
Classic!  Do we need to do a deep copy, meaning that `.slice` doesn't work?

Reply:
Yep, exactly.